### PR TITLE
Master optional products no reload pla

### PR DESCRIPTION
**Pull request by pla-odoo**
Description of the issue/feature this PR addresses:
Task link: https://www.odoo.com/web?debug#id=1961015&action=333&active_id=131&model=project.task&view_type=form&menu_id=4720

Current behavior before PR: 

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_management/static/src/js/sale_management.js [hunk 1]
```diff
@@ -19,28 +19,61 @@ publicWidget.registry.SaleUpdateLineButton = publicWidget.Widget.extend({
     _onClick: function (ev) {
         ev.preventDefault();
         var self = this;
-        var href = this.$el.attr("href");
-        var orderID = href.match(/my\/orders\/([0-9]+)/);
-        var lineID = href.match(/update_line\/([0-9]+)/);
+        var $input = $(ev.currentTarget);
+        this.$input = $input;
```
Comment (qsm-odoo) [inline addons/sale_management/static/src/js/sale_management.js]: Why saving the input variable in `this` ?

File: addons/sale_management/static/src/js/sale_management.js [hunk 2]
```diff
@@ -19,28 +19,61 @@ publicWidget.registry.SaleUpdateLineButton = publicWidget.Widget.extend({
...
         }).then(function (data) {
-            if (!data) {
-                window.location.reload();
+            if (!data['quantity']) {
+                self.$el.find('.sale_order_portal').empty();
+                self.$el.find('.sale_order_portal').append($(data['sale_order_portal_content']).children());
             }
-            self.$el.closest('.input-group').find('.js_quantity').val(data[0]);
-            $('[data-id="total_amount"] > span').html(data[1]);
+            else {
```
Comment (qsm-odoo) [inline addons/sale_management/static/src/js/sale_management.js]: Else on the same line as the closing brace of the if, see https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines

File: addons/sale_management/static/src/js/sale_management.js [hunk 3]
```diff
@@ -19,28 +19,61 @@ publicWidget.registry.SaleUpdateLineButton = publicWidget.Widget.extend({
...
+        var url = "/my/orders/" + parseInt(orderID) + "/update_line";
         this._rpc({
             route: url,
             params: params,
         }).then(function (data) {
-            if (!data) {
-                window.location.reload();
+            if (!data['quantity']) {
+                self.$el.find('.sale_order_portal').empty();
+                self.$el.find('.sale_order_portal').append($(data['sale_order_portal_content']).children());
```
Comment (qsm-odoo) [inline addons/sale_management/static/src/js/sale_management.js]: `self.$el.find(` => `self.$(`

File: addons/sale_management/static/src/js/sale_management.js [hunk 4]
```diff
@@ -19,28 +19,61 @@ publicWidget.registry.SaleUpdateLineButton = publicWidget.Widget.extend({
...
+    events: {
+        'click a.js_add_optional_products': '_onClickOptionalProduct',
+    },
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+    */
```
Comment (qsm-odoo) [inline addons/sale_management/static/src/js/sale_management.js]: Wrong indentation + Missing `@param`

**Comment by pla-odoo** [comment]
> Just some conventions comments, could you fix? Is this task finished according to you? :)

Hello @qsm-odoo changes are done. Please check it!

**Comment by arj-odoo** [comment]
Hello, I have a traceback when I update the optional product quantity with a number (without clicking on the +/- button).

```
2019-11-12 08:47:35,664 26451 INFO test werkzeug: 127.0.0.1 - - [12/Nov/2019 08:47:35] "POST /my/orders/23/update_line_dict HTTP/1.1" 200 - 34 0.094 0.093
2019-11-12 08:47:35,664 26451 ERROR test odoo.sql_db: bad query: UPDATE "sale_order" SET "amount_untaxed"='95542.00',"amount_total"='95542.00',"write_uid"=7,"write_date"=(now() at time zone 'UTC') WHERE id IN (23)
ERROR: could not serialize access due to concurrent update
 
2019-11-12 08:47:35,669 26451 INFO test werkzeug: 127.0.0.1 - - [12/Nov/2019 08:47:35] "POST /my/orders/23/update_line_dict HTTP/1.1" 500 - 32 0.085 0.112
2019-11-12 08:47:35,683 26451 ERROR test werkzeug: Error on request:
Traceback (most recent call last):
  File "/home/arj/PycharmProjects/odoo/odoo_env/lib/python3.6/site-packages/werkzeug/serving.py", line 270, in run_wsgi
    execute(self.server.app)
  File "/home/arj/PycharmProjects/odoo/odoo_env/lib/python3.6/site-packages/werkzeug/serving.py", line 258, in execute
    application_iter = app(environ, start_response)
  File "/home/arj/PycharmProjects/odoo/odoo/service/server.py", line 415, in app
    return self.app(e, s)
  File "/home/arj/PycharmProjects/odoo/odoo/service/wsgi_server.py", line 142, in application
    return application_unproxied(environ, start_response)
  File "/home/arj/PycharmProjects/odoo/odoo/service/wsgi_server.py", line 117, in application_unproxied
    result = odoo.http.root(environ, start_response)
  File "/home/arj/PycharmProjects/odoo/odoo/http.py", line 1294, in __call__
    return self.dispatch(environ, start_response)
  File "/home/arj/PycharmProjects/odoo/odoo/http.py", line 1264, in __call__
    return self.app(environ, start_wrapped)
  File "/home/arj/PycharmProjects/odoo/odoo_env/lib/python3.6/site-packages/werkzeug/wsgi.py", line 766, in __call__
    return self.app(environ, start_response)
  File "/home/arj/PycharmProjects/odoo/odoo/http.py", line 1467, in dispatch
    response = self.get_response(httprequest, result, explicit_session)
  File "/home/arj/PycharmProjects/odoo/odoo/http.py", line 281, in __exit__
    self._cr.commit()
  File "/home/arj/PycharmProjects/odoo/odoo/sql_db.py", line 163, in wrapper
    return f(self, *args, **kwargs)
  File "/home/arj/PycharmProjects/odoo/odoo/sql_db.py", line 388, in commit
    flush_env(self)
  File "/home/arj/PycharmProjects/odoo/odoo/sql_db.py", line 67, in flush_env
    env['base'].flush()
  File "/home/arj/PycharmProjects/odoo/odoo/models.py", line 5344, in flush
    process(self.env[model_name], id_vals)
  File "/home/arj/PycharmProjects/odoo/odoo/models.py", line 5335, in process
    recs._write(vals)
  File "/home/arj/PycharmProjects/odoo/addons/sale/models/sale.py", line 454, in _write
    return super(SaleOrder, self)._write(values)
  File "/home/arj/PycharmProjects/odoo/odoo/models.py", line 3591, in _write
    cr.execute(query, params + [sub_ids])
  File "/home/arj/PycharmProjects/odoo/odoo/sql_db.py", line 163, in wrapper
    return f(self, *args, **kwargs)
  File "/home/arj/PycharmProjects/odoo/odoo/sql_db.py", line 240, in execute
    res = self._obj.execute(query, params)
psycopg2.extensions.TransactionRollbackError: could not serialize access due to concurrent update - - -
```
In the javascript console:

```
http://localhost:8069/my/orders/24/update_line_dict 500 (INTERNAL SERVER ERROR)

```

**Comment by arj-odoo** [comment]
This PR fixes a bug when the quantity of optional quantity was updated with a value (and not the +/- buttons). In master, I get a traceback. 
JS:
```
jquery.js:9600 POST http://localhost:8069/my/orders/21/update_line_dict 500 (INTERNAL SERVER ERROR)
```
Python:
```
bad query: UPDATE "sale_order" SET "amount_untaxed"='74142.00',"amount_total"='74142.00',"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (21)
ERROR: could not serialize access due to concurrent update
psycopg2.extensions.TransactionRollbackError: could not serialize access due to concurrent update 
```

**Comment by jso-odoo** [comment]
@arj-odoo, would you please check it again? I've rebased the branch and tested, I didn't face any traceback.

**Comment by arj-odoo** [comment]
> @arj-odoo, would you please check it again? I've rebased the branch and tested, I didn't face any traceback.

Yes, my bad. I faced the problem in master but your branch seems to fix it.

**Comment by arj-odoo** [comment]
Hello,

There are some variables duplicated in the javascript widget. 
These variables are catched with regex that could be broken if the routes are updated.

The best approach is to update the xml template to add the value you will need with t-att-data fields.

You can save these data in the start function at instantiation, save them in "this" and access them in the methods easily.

Typically, those variable definitions should be modified:
_onChangeQuantity:
var href = $orderLine.find('a.js_update_line_json').attr("href");
var orderID = href.match(/my\/orders\/([0-9]+)/);
var lineID = href.match(/update_line\/([0-9]+)/);
var token = href.match(/token=([\w\d-]*)/)[1];


var href = $orderLine.find('a.js_update_line_json').attr("href");
var orderID = href.match(/my\/orders\/([0-9]+)/);
var lineID = href.match(/update_line\/([0-9]+)/);

_onClick

 var href = $target.attr("href");
 var orderID = href.match(/my\/orders\/([0-9]+)/);
 var lineID = href.match(/update_line\/([0-9]+)/);
 line_id': parseInt(lineID[1]),
 var token = href.match(/token=([\w\d-]*)/)[1];
 
 
_onClickOptionalProduct

 var href = $target.attr("href");
var orderID = href.match(/my\/orders\/([0-9]+)/);
var optionID = href.match(/add_option\/([0-9]+)/);
var token = href.match(/token=([\w\d-]*)/)[1];

Don't hesitate to contact me if you have questions.

**Comment by tivisse** [comment]
Hi @qsm-odoo 

I would like to review this task and I see that you've interacted on the PR. 

Are you ok with those changes or would you like to make another review ? 

Thanks for your input.

**Comment by qsm-odoo** [comment]
@tivisse https://github.com/odoo/odoo/pull/32551/files#diff-2cbf5b7d27b13e90a0622700a29f1965R60 this seems to always be true... anyways, fine by me, I'll let you handle the PR, thanks.

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**